File: 백준/Silver/4921. 나무 블록/나무 블록.py
index = 1

while(True):
  puzzle = input()
  if puzzle == '0':
    break
  
  flag = True
  length = len(puzzle)

  if puzzle[0] != '1' or puzzle[length - 1] != '2':
    flag = False

  if puzzle[length - 2] != '4' and puzzle[length - 2] != '6':
    flag = False

  prev = puzzle[0]
  if flag:
    for item in puzzle[1:length - 1]:
      if item == '1' or item == '2':
        flag = False
        break
      
      if prev == '1':
        if item != '5' and item != '4':
          flag = False
          break

      if prev == '3':
        if item != '5' and item != '4':
          flag = False
          break

      if prev == '4':
        if item != '3' and item != '2':
          flag = False
          break
      
      if prev == '5':
        if item != '8':
          flag = False
          break

      if prev == '6':
        if item != '2' and item != '3':
          flag = False
          break

      if prev == '7':
        if item != '8':
          flag = False
          break

      if prev == '8':
        if item != '7' and item != '6':
          flag = False
          break

      prev = item
    
  if flag:
    print(f"{index}. VALID")

  else:
    print(f"{index}. NOT")

  index += 1

File: 백준/Silver/4921. 나무 블록/나무 블록.c
#include <stdio.h>
#include <string.h>

int main(void) { 
   char puzzle[10001];
   int index = 1;

   while(1){
      scanf("%s", puzzle);
      if(strcmp(puzzle, "0") == 0)   break;

      int flag = 1;
      int len = strlen(puzzle);
      if(puzzle[0] != '1' || puzzle[len - 1] != '2')   flag = 0;
      if(flag){
         for(int i = 0; i < len - 1; i++){
            if((puzzle[i] == '1' && i != 0) || puzzle[i] == '2'){
               flag = 0;
               break;
            }
            
            if(puzzle[i] == '1'){
               if(puzzle[i + 1] != '5' && puzzle[i + 1] != '4'){
                  flag = 0;
                  break;
               }
            }

            if(puzzle[i] == '3'){
               if(puzzle[i + 1] != '5' && puzzle[i + 1] != '4'){
                  flag = 0;
                  break;
               }
            }

            if(puzzle[i] == '4'){
               if(puzzle[i + 1] != '3' && puzzle[i + 1] != '2'){
                  flag = 0;
                  break; 
               }         
            }

            if(puzzle[i] == '5' && puzzle[i + 1] != '8'){
               flag = 0;
               break;
            }

            if(puzzle[i] == '6'){
               if(puzzle[i + 1] != '2' && puzzle[i + 1] != '3'){
                  flag = 0;
                  break;
               }
            }

            if(puzzle[i] == '7' && puzzle[i + 1] != '8'){
               flag = 0;
               break;
            }

            if(puzzle[i] == '8'){
               if(puzzle[i + 1] != '7' && puzzle[i + 1] != '6'){
                  flag = 0;
                  break;
               }
            }
         }
      }
      

      if(flag)   printf("%d. VALID\n", index);
      else   printf("%d. NOT\n", index);

      index++;
   }
   
   return 0;
}